pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1952,18 +1953,23 @@ func buildListener(opts buildListenerOpts) *xdsapi.Listener {
 	listener := &xdsapi.Listener{
 		// TODO: need to sanitize the opts.bind if its a UDS socket, as it could have colons, that envoy
 		// doesn't like
-		Name:            fmt.Sprintf("%s_%d", opts.bind, opts.port),
-		Address:         util.BuildAddress(opts.bind, uint32(opts.port)),
-		ListenerFilters: listenerFilters,
-		FilterChains:    filterChains,
-		DeprecatedV1:    deprecatedV1,
+		Name:             fmt.Sprintf("%s_%d", opts.bind, opts.port),
+		Address:          util.BuildAddress(opts.bind, uint32(opts.port)),
+		ListenerFilters:  listenerFilters,
+		FilterChains:     filterChains,
+		DeprecatedV1:     deprecatedV1,
+		TrafficDirection: opts.trafficDirection,
 	}
 
+	// setup listener filter timeout for sidecar only
 	if util.IsIstioVersionGE13(opts.proxy) && opts.proxy.Type != model.Router {
-		listener.ListenerFiltersTimeout = gogo.DurationToProtoDuration(opts.push.Mesh.ProtocolDetectionTimeout)
+		// This implies the port is explicitly defined as tls-xxx or https-xxx
+		if !(needTLSInspector && !opts.needHTTPInspector && listener.TrafficDirection == core.TrafficDirection_OUTBOUND) {
```

Comment:
nit: this is really complicated, can we break this out into a function or something?

Reply:
Also what if we don't need tls inspector or http inspector? Then seems like we still set timeout?

Reply:
`!(needTLSInspector && !opts.needHTTPInspector && listener.TrafficDirection == core.TrafficDirection_OUTBOUND)` or 
`!needTLSInspector || opts.needHTTPInspector || listener.TrafficDirection != core.TrafficDirection_OUTBOUND`

either is ok.

> Also what if we don't need tls inspector or http inspector? Then seems like we still set timeout?

Yes, we still set timeout, but the timeout is not used since there is no listener filters.

Reply:
probably doesn't matter, but seems safer to not set it if we don't need it

Reply:
Yes, agree. 

I think we need to set the timeout for these two conditions:
1. all inbound listeners
2. outbound listeners with http inspectors

@rshriram @howardjohn could you confirm if I miss something here. If not, I am going to change the condition.